src/main/kotlin/com/exactpro/th2/codec/EventProcessor.kt

```diff
@@ -0,0 +1,209 @@
+        body: List<String>
+    ): String = DEFAULT_EVENT_ID
+
+    override fun storeErrorEvent(
+        message: String,
+        messagesIds: List<MessageID>,
+        cause: Throwable?,
+        additionalBody: List<String>
+    ) = DEFAULT_EVENT_ID
+
+    override fun storeEachEvent(warnEvent: String, message: String, events: Set<EventID>) {}
+    override fun storeEachErrorEvent(errorEventId: String, message: String, events: Set<EventID>) {}
+
+    companion object {
+        private const val DEFAULT_EVENT_ID = "0000"
+    }
+}
+
+class StoreEventProcessor(private val storeEventFunc: (Event, EventID?) -> Unit) : AbstractEventProcessor() {
+
+    override fun storeEvent(
+        message: String,
+        messagesIds: List<MessageID>,
+        body: List<String>
+    ): String {
+        val event = createEvent(message, messagesIds, body = body)
+        storeEventFunc(event, null)
+        return event.id
+    }
+
+    override fun storeErrorEvent(
+        message: String,
+        messagesIds: List<MessageID>,
+        cause: Throwable?,
+        additionalBody: List<String>
+    ): String {
+        val event = createEvent(message, messagesIds, FAILED, cause, additionalBody)
+        storeEventFunc(event, null)
+        return event.id
+    }
+
+    override fun storeEachEvent(
+        warnEvent: String,
+        message: String,
+        events: Set<EventID>
+    ) {
+        events.forEach {
+            it.addReferenceTo(warnEvent, message, PASSED)
+        }
+    }
+
+    override fun storeEachErrorEvent(
+        errorEventId: String,
+        message: String,
+        events: Set<EventID>
+    ) {
+        events.forEach {
+            it.addReferenceTo(errorEventId, message, FAILED)
+        }
+    }
+
+    private fun EventID?.addReferenceTo(eventId: String, name: String, status: Status): EventID {
+        Event.start()
+            .endTimestamp()
+            .name(name)
+            .status(status)
+            .type(if (status != PASSED) "Error" else "Warn")
+            .bodyData(EventUtils.createMessageBean("This event contains reference to the codec event"))
+            .bodyData(ReferenceToEvent(eventId))
+            .also { event ->
+                storeEventFunc(event, this)
+            }.run {
+                return checkNotNull(EventUtils.toEventID(startTimestamp, "", "", id))
+            }
+    }
+
+    private fun createEvent(
+        message: String,
+        messagesIds: List<MessageID> = emptyList(),
+        status: Status = PASSED,
+        cause: Throwable? = null,
+        body: List<String> = emptyList(),
+    ) = Event.start().apply {
+        name(message)
+        type(if (status != PASSED || cause != null) "Error" else "Warn")
+        status(if (cause != null) FAILED else status)
+        messagesIds.forEach(::messageID)
+
+        generateSequence(cause, Throwable::cause).forEach {
+            bodyData(EventUtils.createMessageBean(it.message))
+        }
```

Comment:
There is a method `exception` in `Event` class that allows adding the exception. Let's use that method to avoid code duplication and unify how the exception is displayed in the event

Reply:
Please, update this part as well